plans/objchange/objchange.go

```diff
@@ -102,181 +279,176 @@ func proposedNewObject(schema *configschema.Block, prior, config cty.Value) cty.
-			// should've been caught during an earlier validation step, and
-			// so we don't really care about that here.
-			newV = configV
+			if attr.NestedType != nil {
+				// For non-computed NestedType attributes, we need to descend
+				// into the individual nested attributes to build the final
+				// value, unless the entire nested attribute is unknown.
+				if !configV.IsKnown() {
+					newV = configV
+				} else {
+					newV = proposedNewNestedType(attr.NestedType, priorV, configV)
+				}
+			} else {
+				// For non-computed attributes, we always take the config value,
+				// even if it is null. If it's _required_ then null values
+				// should've been caught during an earlier validation step, and
+				// so we don't really care about that here.
+				newV = configV
+			}
 		}
 		newAttrs[name] = newV
 	}
+	return newAttrs
+}
 
-	// Merging nested blocks is a little more complex, since we need to
-	// correlate blocks between both objects and then recursively propose
-	// a new object for each. The correlation logic depends on the nesting
-	// mode for each block type.
-	for name, blockType := range schema.BlockTypes {
-		priorV := prior.GetAttr(name)
-		configV := config.GetAttr(name)
-		var newV cty.Value
-		switch blockType.Nesting {
+func proposedNewNestedType(schema *configschema.Object, prior, config cty.Value) cty.Value {
+	var newV cty.Value
+	switch schema.Nesting {
+	case configschema.NestingSingle:
+		newAttrs := proposedNewAttributes(schema.Attributes, prior, config)
+		newV = cty.ObjectVal(newAttrs)
+
+	case configschema.NestingList:
+		// Nested blocks are correlated by index.
+		configVLen := 0
+		if config.IsKnown() && !config.IsNull() {
+			configVLen = config.LengthInt()
+		}
+		if configVLen > 0 {
+			newVals := make([]cty.Value, 0, configVLen)
+			for it := config.ElementIterator(); it.Next(); {
+				idx, configEV := it.Element()
+				if prior.IsKnown() && (prior.IsNull() || !prior.HasIndex(idx).True()) {
+					// If there is no corresponding prior element then
+					// we just take the config value as-is.
+					newVals = append(newVals, configEV)
+					continue
+				}
+				priorEV := prior.Index(idx)
 
-		case configschema.NestingSingle, configschema.NestingGroup:
-			newV = ProposedNewObject(&blockType.Block, priorV, configV)
+				newEV := proposedNewAttributes(schema.Attributes, priorEV, configEV)
+				newVals = append(newVals, cty.ObjectVal(newEV))
+			}
+			// Despite the name, a NestingList might also be a tuple, if
+			// its nested schema contains dynamically-typed attributes.
+			if config.Type().IsTupleType() {
+				newV = cty.TupleVal(newVals)
+			} else {
+				newV = cty.ListVal(newVals)
+			}
+		} else {
+			// Despite the name, a NestingList might also be a tuple, if
+			// its nested schema contains dynamically-typed attributes.
+			if config.Type().IsTupleType() {
+				newV = cty.EmptyTupleVal
+			} else {
+				newV = cty.ListValEmpty(schema.ImpliedType())
```

Comment:
I think this is where we're going to want `cty.NullVal(schema.ImpliedType())`. I'm not sure offhand how to handle the tuple case, I'll have to run some tests